source/common/upstream/upstream_impl.cc

```diff
@@ -262,34 +262,37 @@ ResourceManager& ClusterInfoImpl::resourceManager(ResourcePriority priority) con
   return *resource_managers_.managers_[enumToInt(priority)];
 }
 
-void ClusterImplBase::setInitialized() {
-  if (!initialized_ && health_checker_ && pending_initialize_health_checks_ == 0) {
+void ClusterImplBase::startInitialization() {
+  if (!initialization_started_ && health_checker_ && pending_initialize_health_checks_ == 0) {
     pending_initialize_health_checks_ = hosts().size();
-    ASSERT(pending_initialize_health_checks_ > 0);
 
     // TODO(mattklein123): Remove this callback when done.
     health_checker_->addHostCheckCompleteCb([this](HostSharedPtr, bool) -> void {
       if (pending_initialize_health_checks_ > 0 && --pending_initialize_health_checks_ == 0) {
-        setInitialized();
+        finishInitialization();
       }
     });
   }
 
-  initialized_ = true;
-  if (initialize_callback_ && pending_initialize_health_checks_ == 0) {
-    // Every time a host changes HC state we cause a full healthy host recalculation which
-    // for expensive LBs (ring, subset, etc.) can be quite time consuming. During startup, this
-    // can also block worker threads by doing this repeatedly. There is no reason to do this
-    // as we will not start taking traffic until we are initialized. By blocking HC updates
-    // while initializing we can avoid this. When HC responses for all hosts have arrived and
-    // we are about to initialize, we unblock further HC updates which has the additional effect
-    // of forcing a healthy host recalculation.
+  if (!initialization_started_ && pending_initialize_health_checks_ == 0) {
+    finishInitialization();
+  }
+
+  initialization_started_ = true;
+}
+
+void ClusterImplBase::finishInitialization() {
+  // Snap a copy of the completion callback so that we can set it to nullptr to unblock
+  // reloadHealthyHosts(). See that function for more info on why we do this.
+  auto snapped_callback = initialization_complete_callback_;
+  initialization_complete_callback_ = nullptr;
+
+  if (snapped_callback != nullptr) {
     if (health_checker_ != nullptr) {
       reloadHealthyHosts();
```

Comment:
Ok, now that these are split out it's a bit weird to me that we'd only reloadHealthyHosts if we had an initialize callback set.  

I'd say at worst, let's either comment this up in the .h to make it really clear, or maybe add a dummy closure on creation so there's always a callback until initialization is complete.  I think it'd be cleanest to make startInitialize() take the callback to be called on completion so it's guaranteed to work "as expected" but that may be more yaks than you want to shave today :-)

Reply:
This is actually to work around a different issue. I will add a comment.